Copter: chase/follow mode (attempt 2)

rmackay9 opened the pull request:
#This PR builds on this earlier PR from @peterbarker (https://github.com/ArduPilot/ardupilot/pull/7460) to add "Chase" mode which allows a vehicle to follow another vehicle or ground station.

- chase mode is built on Guided mode's velocity controller to fly towards a vehicle's position provided by the AP_Follow library.  The control uses the other vehicle's velocity as a base and then adds a correction to try and reduce the position error.
- AP_Follow library is responsible for:
  - consuming mavlink messages from the target vehicle (currently only GLOBAL_POSITION_INT is supported)
  - providing a 3D distance vector to the other vehicle both with and without "offsets" (both are required to allow moving to a specified distance from the other vehicle while remaining pointed at it) and a 3D velocity vector
  - FOLL_DIST_MAX holds the maximum distance to the other vehicle, beyond this the vehicle is ignored
  - FOLL_SYSID should be set to the other vehicle's mavlink id or left as zero to track the first vehicle (or ground station) that provides a GLOBAL_POSITION_INT message
  - FOLL_OFS_X, Y, Z allows specifying an offset from the lead vehicle that we should try and maintain.  Leaving these as zero means the offset will be initialised to the distance between the vehicles when the vehicle is switched into Chase mode.
  - FOLL_OFS_TYPE allows specifying if the above offset should be relative to the lead vehicle's heading or in NED frame.
  - FOLL_YAW_BEHAVE specifies if this vehicle should point at the lead vehicle or in the same direction as the lead vehicle

Things to address is follow up PRs:

- add support for the mavlink FOLLOW_TARGET (http://mavlink.org/messages/common#FOLLOW_TARGET) message that includes accelerations
- add support for DO_FOLLOW_REPOSITION to support setting offsets
- add support for NAV_FOLLOW to switch into chase/follow mode (?)
- support pointing the camera at the lead vehicle
- add filtering to smooth the response
- add avoidance of the lead vehicle so we don't accidentally hit it which could happen if the offsets lead to this vehicle trying to fly from it's current position to the opposite side of the lead vehicle
- allow the user to change the offset position from the lead vehicle using the transmitter (throttle for altitude, roll and pitch to change horizontal position)

Pedals2Paddles commented:
What do you think about using the RC inputs to adjust the offsets on the fly? Throttle to adjust altitude offset, pitch and roll for lateral. This would let you use stalker mode without having to constantly send mavlink commands to adjust position.

rmackay9 commented:
@Pedals2Paddles, yes, @peterbarker has brought up this idea as well.  handling altitude won't be too difficult.  Making the horizontal movements feel natural could be tricky but still a good thing to do so I've added it to the list of items to address in future PRs above.

amilcarlucas commented on libraries/AP_Follow/AP_Follow.cpp:
```diff
@@ -0,0 +1,281 @@
+        mavlink_msg_global_position_int_decode(&msg, &packet);
+        _target_location.lat = packet.lat;
+        _target_location.lng = packet.lon;
+        _target_location.alt = packet.alt/10;       // convert millimeters to cm
```
Use `* 0.1` instead?

amilcarlucas commented on libraries/AP_Follow/AP_Follow.cpp:
```diff
@@ -0,0 +1,281 @@
+        _target_location.lat = packet.lat;
+        _target_location.lng = packet.lon;
+        _target_location.alt = packet.alt/10;       // convert millimeters to cm
+        _target_velocity_ned.x = packet.vx/100.0f;  // velocity north
```
Use `* 0.01f` instead?

amilcarlucas commented on libraries/AP_Follow/AP_Follow.cpp:
```diff
@@ -0,0 +1,281 @@
+        _target_location.lng = packet.lon;
+        _target_location.alt = packet.alt/10;       // convert millimeters to cm
+        _target_velocity_ned.x = packet.vx/100.0f;  // velocity north
+        _target_velocity_ned.y = packet.vy/100.0f;  // velocity east
```
Use `* 0.01f` instead?

amilcarlucas commented on libraries/AP_Follow/AP_Follow.cpp:
```diff
@@ -0,0 +1,281 @@
+        _target_location.alt = packet.alt/10;       // convert millimeters to cm
+        _target_velocity_ned.x = packet.vx/100.0f;  // velocity north
+        _target_velocity_ned.y = packet.vy/100.0f;  // velocity east
+        _target_velocity_ned.z = packet.vz/100.0f;  // velocity down
```
Use `* 0.01f` instead?

amilcarlucas commented on libraries/AP_Follow/AP_Follow.cpp:
```diff
@@ -0,0 +1,281 @@
+        _target_velocity_ned.z = packet.vz/100.0f;  // velocity down
+        _last_location_update_ms = now;
+        if (packet.hdg <= 36000) {                  // heading (UINT16_MAX if unknown)
+            _target_heading = packet.hdg / 100.0f;  // convert centi-degrees to degrees
```
Use `* 0.01f` instead?

rmackay9 replied:
@amilcarlucas, thanks for the review, fixed!

Pedals2Paddles commented:
At what rate do you suggest the vehicle being followed transmit the `global_position_int` message?  What do you think about some kind of straight forward way for the operator the turn on/off the transmission of the `global_position_int` message?  Maybe an RC CH option?  Or a mavlink command to start/stop it?

rmackay9 commented:
@Pedals2Paddles, I'm not sure what rate will be required but I'm sure faster than 1hz will be required at least if the vehicle is changing directions a lot.  If the update rate is too slow we will see a jerky response in the following vehicle.  I think at 10hz the results should be quite good.

To stop a vehicle from following, I think we should somehow send a COMMAND_LONG or COMMAND_LONG_INT command with a DO_FOLLOW message in it with the first field (for sysid) set to zero.  The thing is that the lead vehicle doesn't really know it's being followed.  At least with the current implementation.

Pedals2Paddles commented:
If configured for high / neutral / low, a single RC channel switch set high can initiate transmitting position and transmit the DO_FOLLOW command.  Continue transmitting position while neutral.  And cease transmitting position, sending the stop stalking me (0) command when set slow.  That would allow a single operator of the vehicle being stalked to initiate and halt the stalkers.  Without needing to stop flying and go to the GCS.

khancyr commented on ArduCopter/mode_follow.cpp:
```diff
@@ -0,0 +1,155 @@
+            }
+
+            case AP_Follow::YAW_BEHAVE_SAME_AS_LEAD_VEHICLE: {
+                float target_hdg = 0.0f;;
```
double comma

rmackay9 replied:
txs, fixed!

khancyr commented on libraries/AP_Follow/AP_Follow.cpp:
```diff
@@ -0,0 +1,281 @@
+    }
+
+    // calculate time since last actual position update
+    float dt = (AP_HAL::millis() - _last_location_update_ms) * 0.001f;
```
const

rmackay9 replied:
txs, added const

khancyr commented on libraries/AP_Follow/AP_Follow.cpp:
```diff
@@ -0,0 +1,281 @@
+    }
+
+    // calculate difference
+    Vector3f dist_vec = location_3d_diff_NED(current_loc, target_loc);
```
const

rmackay9 replied:
txs, added const here too.

Pedals2Paddles commented on ArduCopter/mode_follow.cpp:
```diff
@@ -0,0 +1,155 @@
+#define Debug(fmt, args ...)
+#endif
+
+// initialise avoid_adsb controller
```
Presumably a copy/paste thing?

rmackay9 replied:
ah, thanks for that catch, fixed!

Pedals2Paddles commented:
If `FOLL__YAW_BEHAVE` is set to none, therefore `use_yaw` remains false, therefore no guided mode `set_auto_yaw_look_at_heading` is called, will the RC yaw continue to function under pilot control?

rmackay9 commented:
I flight tested this extensively yesterday and have found a couple of issues that I've corrected with follow up commits.  Performance was not very good but I think the issue is the 900mHz radios I'm using are capable of less than 1hz update rates.

I've added the WIP label until I can get the performance acceptable.

amilcarlucas commented on ArduCopter/GCS_Mavlink.cpp:
```diff
@@ -134,28 +134,22 @@ NOINLINE void Copter::send_extended_status1(mavlink_channel_t chan)
-        fix_time = gps.last_fix_time_ms();
-    } else {
-        fix_time = millis();
+    const uint32_t now = AP_HAL::millis();
```
Can we ask the EKF when it calculated the position and use that here instead ? That would inprove the jitter that I was talking about in the dev meeting.

rmackay9 replied:
@amilcarlucas, before I made this change to use system time the position reported could be up to 200ms off but after this change it should be down to a max of 2.5ms.  I suspect the transmission times of most radios will drown out this difference... no?

amilcarlucas replied:
Yes... but for high speed, RTK systems those 2.5 ms might get interesting. How hard is to get them ?

rmackay9 replied:
@amilcarlucas, I'm sure it's possible but I'm not sure how difficult it would be.  We still have a ways to go before the this level of lag the biggest issue i think.

rmackay9 commented:
@Pedals2Paddles, some how I failed to respond to your question above.. sorry about that.  So at the moment the follow mode doesn't accept any user input from the sticks at all.  That could be added but I thought we could do that in a follow up PR.

amilcarlucas commented on libraries/AP_Follow/AP_Follow.cpp:
```diff
@@ -0,0 +1,336 @@
+}
+
+// get distance vector to target (in meters) and target's velocity all in NED frame
+bool AP_Follow::get_target_dist_and_vel_ned(Vector3f &dist, Vector3f &dist_with_offs, Vector3f &vel)
```
And please also add unit suffixes, the crazy mix of `m` and `cm` makes the code difficult to follow. Explicit units is a good thing !!!

rmackay9 replied:
i definitely recognise the mixing of units is not good... paul riseborough mentioned this more generally at the un-conference.  It comes from the deeper problem of having the underlying libraries also use cm.

amilcarlucas commented on libraries/AP_Follow/AP_Follow.cpp:
```diff
@@ -0,0 +1,336 @@
+        _target_velocity_ned.x = packet.vx * 0.01f; // velocity north
+        _target_velocity_ned.y = packet.vy * 0.01f; // velocity east
+        _target_velocity_ned.z = packet.vz * 0.01f; // velocity down
+        _last_location_update_ms = now;
```
It would be a LOT better if we used the source timestamp of the message instead of now :)

rmackay9 replied:
Yes, although I think we need to store both times.  At the moment _last_location_update_ms is used for two purposes:

1. timeout calcs if we haven't received a message in a while
2. calculating the dt to extrapolate the current position from the last update message

the timeout should use our system time while the dt could use the system time of the other vehicle received in the message.

I think I'll add this as an item on the to-do list so we don't forget about it.

amilcarlucas commented on libraries/AP_Follow/AP_Follow.cpp:
```diff
@@ -0,0 +1,336 @@
+        _last_location_update_ms = now;
+        if (packet.hdg <= 36000) {                  // heading (UINT16_MAX if unknown)
+            _target_heading = packet.hdg * 0.01f;   // convert centi-degrees to degrees
+            _last_heading_update_ms = now;
```
Again source timestamp. For that we need to know when the other system booted, relative to our own boot time.
But ... guess what ... GPS can give us that :) an absolute time reference common to all UAVs and/or GCSs

rmackay9 replied:
Actually I think in this case, we are only using this time for the timeout so using our system clock is OK i think.

amilcarlucas commented on libraries/AP_Follow/AP_Follow.cpp:
```diff
@@ -0,0 +1,336 @@
+                            (unsigned)_sysid_to_follow,
+                            (long)_target_location.lat,
+                            (long)_target_location.lng,
+                            (double)(_target_location.alt * 0.01f));
```
add `// cm -> m`

amilcarlucas replied:
should'nt it be float instead of double ???

rmackay9 replied:
whenever we use send_text it apparently needs to be a double.  a compiler warning pops up if we try to use float.

amilcarlucas commented on libraries/AP_Follow/AP_Follow.cpp:
```diff
@@ -0,0 +1,336 @@
+        }
+
+        // log lead's estimated vs reported position
+        DataFlash_Class::instance()->Log_Write("FOLL",
```
The units and multiplication factors are missing

rmackay9 replied:
Ok, added!

amilcarlucas commented on libraries/AP_Follow/AP_Follow.cpp:
```diff
@@ -0,0 +1,336 @@
+}
+
+// get velocity estimate in NED frame
+bool AP_Follow::get_velocity_ned(Vector3f& vel, float dt) const
```
replace `vel` with `vel_ned_ms`

rmackay9 replied:
I think normally we leave out the trailing units indication if it is already in SI.  So we should add the trailing units to the name when it's cm but not when it's meters.

amilcarlucas commented on libraries/AP_Follow/AP_Follow.cpp:
```diff
@@ -0,0 +1,336 @@
+}
+
+// initialise offsets to target if required
+void AP_Follow::init_offsets_if_required(const Vector3f& dist_vec_ned)
```
Is this in m or cm ??

rmackay9 replied:
it's in meters.  thanks, fixed in follow up commit.

amilcarlucas commented on libraries/AP_Follow/AP_Follow.cpp:
```diff
@@ -0,0 +1,336 @@
+}
+
+// get offsets in NED frame
+bool AP_Follow::get_offsets_ned(Vector3f& offset) const
```
Again, please add units

amilcarlucas commented on libraries/AP_Follow/AP_Follow.h:
```diff
@@ -0,0 +1,114 @@
+public:
+
+    // enum for YAW_BEHAVE parameter
+    enum YawBehave {
```
This enum should be pushed to mavlink

amilcarlucas commented on libraries/AP_Follow/AP_Follow.h:
```diff
@@ -0,0 +1,114 @@
+    YawBehave get_yaw_behave() const { return (YawBehave)_yaw_behave.get(); }
+
+    // get target's heading in degrees (0 = north, 90 = east)
+    bool get_target_heading(float &heading) const;
```
rename to `get_target_heading_deg`

amilcarlucas commented on libraries/AP_Follow/AP_Follow.h:
```diff
@@ -0,0 +1,114 @@
+private:
+
+    // get velocity estimate in NED frame using dt since last update
+    bool get_velocity_ned(Vector3f& vel, float dt) const;
```
it is missing units

rmackay9 replied:
added units to comments

amilcarlucas commented on libraries/AP_Follow/AP_Follow.h:
```diff
@@ -0,0 +1,114 @@
+    bool get_velocity_ned(Vector3f& vel, float dt) const;
+
+    // initialise offsets to provided distance vector if required
+    void init_offsets_if_required(const Vector3f& dist_vec_ned);
```
It is missing units information

amilcarlucas commented on libraries/AP_Follow/AP_Follow.h:
```diff
@@ -0,0 +1,114 @@
+    void init_offsets_if_required(const Vector3f& dist_vec_ned);
+
+    // get offsets in NED frame
+    bool get_offsets_ned(Vector3f& offsets) const;
```
It is missing units information

amilcarlucas commented on libraries/AP_Follow/AP_Follow.h:
```diff
@@ -0,0 +1,114 @@
+    AP_Float    _dist_max;          // maximum distance to target.  targets further than this will be ignored
+    AP_Int8     _offset_type;       // offset frame type (0:North-East-Down, 1:RelativeToLeadVehicleHeading)
+    AP_Vector3f _offset;            // offset from lead vehicle in meters
+    AP_Int8     _yaw_behave;        // following vehicle's yaw/heading behaviour
```
add `see YAW_BEHAVE enum`

amilcarlucas commented on libraries/AP_Follow/AP_Follow.h:
```diff
@@ -0,0 +1,114 @@
+    Vector3f _target_velocity_ned;  // last known velocity of target in NED frame in m/s
+    Vector3f _target_accel_ned;     // last known acceleration of target in NED frame in m/s/s
+    uint32_t _last_heading_update_ms;   // system time of last heading update
+    float _target_heading;          // heading in degrees
```
please add _deg sufix

amilcarlucas commented on libraries/AP_Follow/AP_Follow.h:
```diff
@@ -0,0 +1,114 @@
+    bool _healthy;                  // true if we are receiving mavlink messages (regardless of whether they have target position info within them)
+    uint32_t _last_location_update_ms;  // system time of last position update
+    Location _target_location;      // last known location of target
+    Vector3f _target_velocity_ned;  // last known velocity of target in NED frame in m/s
```
please add _ms suffix

amilcarlucas commented on libraries/AP_Follow/AP_Follow.h:
```diff
@@ -0,0 +1,114 @@
+    uint32_t _last_location_update_ms;  // system time of last position update
+    Location _target_location;      // last known location of target
+    Vector3f _target_velocity_ned;  // last known velocity of target in NED frame in m/s
+    Vector3f _target_accel_ned;     // last known acceleration of target in NED frame in m/s/s
```
please add `_mss` suffix

amilcarlucas commented on libraries/AP_Follow/AP_Follow.h:
```diff
@@ -0,0 +1,114 @@
+
+    uint32_t _last_location_sent_to_gcs; // last time GCS was told position
+
+    uint8_t _sysid_to_follow = 0;
```
Add // Mavlink ID of system to follow

rmackay9 replied:
added comments to sysid-to-follow

amilcarlucas commented:
@rmackay9 @peterbarker please take a look at rmackay9/rmackay9-ardupilot#15 before looking at the comments I posted above.

rmackay9 commented on ArduCopter/mode_follow.cpp:
```diff
@@ -0,0 +1,170 @@
+        // limit desired velocity to be between maximum climb and descent rates
+        desired_velocity.z = constrain_float(desired_velocity.z, -fabsf(pos_control->get_speed_down()), pos_control->get_speed_up());
+
+        // unit vector towards target position
```
ok, added to comments in follow up commit.

rmackay9 commented on ArduCopter/mode_follow.cpp:
```diff
@@ -0,0 +1,170 @@
+
+        // slow down horizontally as we approach target (use 1/2 of maximum deceleration for gentle slow down)
+        const float dist_to_target_xy = Vector2f(dist_vec_offs_neu.x, dist_vec_offs_neu.y).length();
+        copter.avoid.limit_velocity(pos_control->get_pos_xy_p().kP().get(), pos_control->get_accel_xy() / 2.0f, desired_velocity_xy, dir_to_target_xy, dist_to_target_xy, copter.G_Dt);
```
txs. changed to use 0.5f in follow up commit.

rmackay9 commented on ArduCopter/mode_follow.cpp:
```diff
@@ -0,0 +1,170 @@
+        desired_velocity.y = desired_velocity_xy.y;
+
+        // limit vertical desired_velocity to slow as we approach target (we use 1/2 of maximum deceleration for gentle slow down)
+        const float des_vel_z_max = copter.avoid.get_max_speed(pos_control->get_pos_z_p().kP().get(), pos_control->get_accel_z() / 2.0f, fabsf(dist_vec_offs_neu.z), copter.G_Dt);
```
fixed in follow up commit.

rmackay9 commented on ArduCopter/mode_follow.cpp:
```diff
@@ -0,0 +1,170 @@
+
+        // limit vertical desired_velocity to slow as we approach target (we use 1/2 of maximum deceleration for gentle slow down)
+        const float des_vel_z_max = copter.avoid.get_max_speed(pos_control->get_pos_z_p().kP().get(), pos_control->get_accel_z() / 2.0f, fabsf(dist_vec_offs_neu.z), copter.G_Dt);
+        desired_velocity.z = constrain_float(desired_velocity.z, -des_vel_z_max, des_vel_z_max);
```
yes, we have separate up and down speed limits and those are applied higher in the mode_follow.cpp.

// limit desired velocity to be between maximum climb and descent rates
desired_velocity.z = constrain_float(desired_velocity.z, -fabsf(pos_control->get_speed_down()), pos_control->get_speed_up());

rmackay9 commented on ArduCopter/mode_follow.cpp:
```diff
@@ -0,0 +1,170 @@
+/*
+ * mode_follow.cpp - follow another mavlink-enabled vehicle by system id
+ *
+ * TODO: set ROI yaw mode / point camera at target
```
txs, removed TODO

rmackay9 commented on libraries/AC_Avoidance/AC_Avoid.cpp:
```diff
@@ -229,6 +229,37 @@ void AC_Avoid::adjust_roll_pitch(float &roll, float &pitch, float veh_angle_max)
+ * Uses velocity adjustment idea from Randy's second email on this thread:
+ * https://groups.google.com/forum/#!searchin/drones-discuss/obstacle/drones-discuss/QwUXz__WuqY/qo3G8iTLSJAJ
+ */
+void AC_Avoid::limit_velocity(float kP, float accel_cmss, Vector2f &desired_vel, const Vector2f& limit_direction, float limit_distance, float dt) const
```
I think I'll leave out the renaming to add _cm and _cms everywhere.  It's a pretty general problem all through our code so.. not sure I'm up to trying to resolve it just yet.

amilcarlucas replied:
You do not nedd to rename it everywhere. I already did that for you in my PR to your personal repository.

Fact is: the current codebase has crazy units and it is not clear on their name what units there is. This saves time, and makes it clear.

rmackay9 commented on libraries/AC_Avoidance/AC_Avoid.h:
```diff
@@ -62,6 +62,19 @@ class AC_Avoid {
+    // limit_direction to be at most the maximum speed permitted by the limit_distance.
+    // uses velocity adjustment idea from Randy's second email on this thread:
+    //   https://groups.google.com/forum/#!searchin/drones-discuss/obstacle/drones-discuss/QwUXz__WuqY/qo3G8iTLSJAJ
+    void limit_velocity(float kP, float accel_cmss, Vector2f &desired_vel, const Vector2f& limit_direction, float limit_distance, float dt) const;
```
Yes, limit_direction must be a unit vector.  I think all our calls to this to pass in a unit vector although i suppose we could add some defensive checks..

amilcarlucas replied:
At least add a WARNING in the comments

rmackay9 commented on libraries/AP_Follow/AP_Follow.cpp:
```diff
@@ -0,0 +1,336 @@
+
+    // @Param: _OFS_Y
+    // @DisplayName: Follow offsets in meters east/right
+    // @Description: Follow offsets in meters east/right.  If positive, this vehicle fly to the right or east of lead vehicle
```
txs, added to param description

rmackay9 commented on libraries/AP_Follow/AP_Follow.cpp:
```diff
@@ -0,0 +1,336 @@
+    // @Param: _YAW_BEHAVE
+    // @DisplayName: Follow yaw behaviour
+    // @Description: Follow yaw behaviour
+    // @Values: 0:None,1:Face Lead Vehicle,2:Same as Lead vehicle,3:Direction of Flight
```
not sure about adding the enum to mavlink because the parameter's value is not currently sent within any mavlink messages

rmackay9 commented on libraries/AP_Follow/AP_Follow.cpp:
```diff
@@ -0,0 +1,336 @@
+}
+
+// get target's estimated location
+bool AP_Follow::get_target_location_and_velocity(Location &loc, Vector3f& vel) const
```
ok, renamed arguments to vel_ned

rmackay9 commented:
I've addressed most of the issues raised, rebased on master and squashed the commits together so I think/hope this is ready to go into master but if there are any further issues I'm happy to address them.

Note: there are some valid issues raised (especially from @amilcarlucas) regarding variable naming and units (cm vs meters) which I hope we can address in follow up PRs.

amilcarlucas commented:
`#if MODE_FOLLOW_ENABLED == ENABLED` is still missing in many places.

rmackay9 commented:
@amilcarlucas, txs for the comments, I've fixed up the build when follow is excluded.

I plan to merge this later today.  I understand that it's not perfect but it shouldn't negatively affect other features/modes of ArduPilot and I think we shouldn't allow perfect to be the enemy of the good.

amilcarlucas commented:
@rmackay9 I understand that we should not allow "the perfect to be the enemy of the good" but...
I think most commits should have been further squashed, their commit messages bring almost no value and  the text in the description of this PR should have been used in the commit message.
That would have provided a much better git history !!!!!!

peterbarker commented:
On Mon, 5 Mar 2018, Amilcar Lucas wrote:

> Fact is: the current codebase has crazy units and it is not clear on their
> name what units there is. This saves time, and makes it clear.

Yes, but I'll say again I think we should only label crazy units, not sane 
units.

amilcarlucas commented:
Peter not a good idea. The code is full of crazy unlabeled units. Creating new code that does not label units, just perpetuates the problem: You never know for sure which unit it is. You always need to double check:
* Is it new sane code with normal units ?
* is it old code with crazy unlabeled units ?

And BTW what is sane `deg` or `rad` ???

Sorry about this but I still think that explicit units is the pragmatic only solution.
It will take years to clean all the codebase. It makes more sense to no longer produce ambiguous code.

amilcarlucas commented:
I rebased https://github.com/rmackay9/rmackay9-ardupilot/pull/15 into master at #7845
